examples/cpp/dds/CustomPayloadPoolExample/CustomPayloadPoolDataPublisher.cpp

```diff
@@ -0,0 +1,210 @@
+        payload_pool_);
+
+    if (writer_ == nullptr)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+CustomPayloadPoolDataPublisher::~CustomPayloadPoolDataPublisher()
+{
+    if (writer_ != nullptr)
+    {
+        publisher_->delete_datawriter(writer_);
+    }
+    if (publisher_ != nullptr)
+    {
+        participant_->delete_publisher(publisher_);
+    }
+    if (topic_ != nullptr)
+    {
+        participant_->delete_topic(topic_);
+    }
+    DomainParticipantFactory::get_instance()->delete_participant(participant_);
+}
+
+void CustomPayloadPoolDataPublisher::on_publication_matched(
+        eprosima::fastdds::dds::DataWriter*,
+        const eprosima::fastdds::dds::PublicationMatchedStatus& info)
+{
+    if (info.current_count_change == 1)
+    {
+        matched_ = info.total_count;
+        first_connected_ = true;
+        std::cout << "Publisher matched." << std::endl;
+    }
+    else if (info.current_count_change == -1)
+    {
+        matched_ = info.total_count;
+        std::cout << "Publisher unmatched." << std::endl;
+    }
+    else
+    {
+        std::cout << info.current_count_change
+                  << " is not a valid value for PublicationMatchedStatus current count change" << std::endl;
+    }
+}
+
+void CustomPayloadPoolDataPublisher::run_thread(
+        uint32_t samples,
+        uint32_t sleep)
+{
+    if (samples == 0)
+    {
+        while (!stop_)
+        {
+            if (publish(false))
+            {
+                std::cout << "Message: " << hello_.message() << " with index: " << hello_.index()
+                          << " SENT" << std::endl;
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(sleep));
+        }
+    }
+    else
+    {
+        for (uint32_t i = 0; i < samples; ++i)
+        {
+            if (!publish())
+            {
+                --i;
+            }
+            else
+            {
+                std::cout << "Message: " << hello_.message() << " with index: " << hello_.index()
+                          << " SENT" << std::endl;
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(sleep));
+        }
+    }
+}
+
+void CustomPayloadPoolDataPublisher::run(
+        uint32_t samples,
+        uint32_t sleep)
+{
+    stop_ = false;
+    std::thread thread(&CustomPayloadPoolDataPublisher::run_thread, this, samples, sleep);
+    if (samples == 0)
+    {
+        std::cout << "Publisher running. Please press enter to stop the Publisher at any time." << std::endl;
+        std::cin.ignore();
+        stop_ = true;
+    }
+    else
+    {
+        std::cout << "Publisher running " << samples << " samples." << std::endl;
+    }
+    thread.join();
+}
+
+bool CustomPayloadPoolDataPublisher::publish(
+        bool wait_for_listener)
+{
+    if (first_connected_ || !wait_for_listener || matched_ > 0)
+    {
+        hello_.index(hello_.index() + 1);
+        writer_->write(&hello_);
+        return true;
```

Comment:
We should return based on the `write` return

Reply:
I still think we should return false in the run function if the write method fails. This way, we can have a return code on the application